sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergUtils.java

```diff
@@ -282,12 +306,50 @@ private static void copyFieldIntoRecord(Record rec, Types.NestedField field, Row
         Optional.ofNullable(value.getDouble(name)).ifPresent(v -> rec.setField(name, v));
         break;
       case DATE:
-        throw new UnsupportedOperationException("Date fields not yet supported");
+        Optional.ofNullable(value.getLogicalTypeValue(name, LocalDate.class))
+            .ifPresent(v -> rec.setField(name, v));
+        break;
       case TIME:
-        throw new UnsupportedOperationException("Time fields not yet supported");
+        Optional.ofNullable(value.getLogicalTypeValue(name, LocalTime.class))
+            .ifPresent(v -> rec.setField(name, v));
+        break;
       case TIMESTAMP:
-        Optional.ofNullable(value.getDateTime(name))
-            .ifPresent(v -> rec.setField(name, v.getMillis()));
+        Object val = value.getValue(name);
+        if (val == null) {
+          break;
+        }
+        Types.TimestampType ts = (Types.TimestampType) field.type().asPrimitiveType();
+        // timestamp with timezone
+        if (ts.shouldAdjustToUTC()) {
+          // currently only string is supported because other types
+          // do not maintain timezone information
+          if (val instanceof String) {
+            // e.g. 2007-12-03T10:15:30+01:00
+            rec.setField(name, OffsetDateTime.parse((String) val));
+            break;
+          } else {
+            throw new UnsupportedOperationException(
```

Comment:
As you mentioned already [here](https://github.com/apache/beam/pull/32688#issuecomment-2402624470), I'd also like to have the option to convert `Instant` and `LocalDateTime` into `OffsetDateTime` at UTC. It will be a lossless conversion and users won't be forced to use Strings.

Reply:
I refactored the conversion logic so that now, any of these types can write to both `timestamp` and `timestamptz`:
- Schema.FieldType.INT64         --> long
- Schema.FieldType.INT64.        --> String
- Schema.FieldType.DATETIME  --> org.joda.time.DateTime
- SqlTypes.DATETIME                  --> java.time.LocalDateTime 

When writing to `timestamptz`, we will resolve to UTC first before writing. Lmk how things look now. I updated the java doc as well with more details

Reply:
And when reading, these are the mappings:
- `timestamp` --> SqlTypes.DATETIME
- `timestamptz` --> Schema.FieldType.DATETIME

Reply:
Yes, thank you. It makes a lot of sense to me 👍

Reply:
LGTM, I left few minor comments, Thank you